### ISSUE TITLE
Entities is weird used as a superclass for Model, Groups, Components, etc.

### ISSUE THREAD

**Issue by DanRathbun** [stevensrmiller/SketchUpFactory#3]
### Entities is weird used as a superclass for Model, Groups, Components.

The Entities collection class is a safe wrapper around C++ collections, whose members are Drawingelement subclass objects, being subclass of Entity. (Entity subclass objects are things that are saved into the model DOM.)

There are **only 2** class objects that have (own) an Entities collection. The **model itself** (which becomes a component when saved to a `.skp` file,) and **ComponentDefinition** objects.

There are 3 kinds of definitions (having differing behavior): normal, image and group definitions. 
(It seems like the C API does not want anyone messing with image definitions as they are fragile, and can cause SketchUp to crash if malformed. A lot of users will immediately wrap them in a group instance, and then explode the image instance to it's materialed face and use the group.)

Group and Image definitions are hidden from the "In Model" components collection in the GUI Components inspector panel of the SketchUp application.

ComponentInstances are the instances of normal definitions (whose `image?` and `group?` flags are false.)
Image objects are the instances of definitions whose `image?` flag is set.
Group instances are the instances of definitions whose `group?` flag is set. 

When the user in SketchUp double-clicks a group instance to enter it's editing context (it is actually editing it's definition's entities collection.) But there is catch that only occurs in the GUI application. If the group's definition has more than 1 instance, then a new clone definition for the group instance entered, is created and the user begins editing the new group definition. 
The cloned definition's Name is appended with a pound sign and an unique integer. (See the issue #1.)
(This is an attempt to fool the userbase into thinking that groups have only 1 instance and components can have many. But for the extension developer, it is powerful to have instances that are hidden from the "In Model" Component library where users could grab, drag and drop plugin components in the wrong way and mess up the plugin or the model. In fact, SketchUp's native Dynamic Component extension uses nested dynamic groups that need to be hidden from user's eyes for several good reasons.)

The APIs (both Ruby and C) try to hide the fact that groups are really special component instances. They provide a way of thinking that groups have Entities collections, when their getters are really shortcut wrappers to their definition's getters.
`group.entities == group.definition.entities`
returns `true`. Ie, the same collection object belonging to the definition of the group.

It was not until the v2015 release that coders got a reliable `group.definition` getter method. Before that release the Ruby API tried to restrict code from directly accessing the group's definition object. But there were many reasons to access it, and iterations of the model's DefinitionList collection often returned the incorrect definition object (because of an API comparison method error, since fixed.)
Allowing coders access simplified code that operated upon all three (component instances, groups and the model. Before this code had to have complex `if ... elsif ... elsif` statements duck-typing the object or checking it's class.

**Comment by DanRathbun** [comment]
Okay, I know your goal was to **simplify things** for simple Unity models.
I get it. It'll likely work for standalone simple models.

But it's just weird for us who know how things are really organized.

This will not work for the "live" model. The main reason is that (as explained above) in the manual edit process, the user enters an editing context, which is an Entities class instance collection belonging to the definition of the group or instance that was double-clicked.

This called the **active_entities** context. (ie, the entities collection being edited.) The core SketchUp C++ code is written to operate on this context, and it's functions (and therefore the exposure of them to the "live" APIs,) also must take this into account. One of the things that devs have to deal with is that the transformation (coordinate system) changes as the active editing context changes.

When in the toplevel (model) entities, all the functions use local coordinates. Code can edit a definition's entities specifying coordinates in the definition's local coordinates. But as the user drills down into instances, the core combines the transforms of each successive context, meaning the core functions change to using global coordinates. Since these functions are exposed this way, an API extension coder has to adapt his code to handle the change.

You won't see this or need to worry when working on standalone SKP files, because there is no user changing the edit context on you.

----

The other main thing is that it is the OOP API collection class objects that control access to and create the collection members via factory methods. Your library here hides the model collections.
 ( More on this in another issue. See: [#4: Entity subclasses have public constructors](https://github.com/stevensrmiller/SketchUpFactory/issues/4) )

**Comment by stevensrmiller** [comment]
I wrestled with subclassing Entities quite a bit. SInce Models / Definitions / Groups all use the "has a" relationship to Entities (and we are supposed to favor composition over inheritance), that seemed the way to do it, at first. (I note your detailed break-down of Groups distinguishes them from the other two, but that's an implementation detail that is not apparent from the perspective of client code solely using the C API.)

But...

I wanted client code to be able to add polygons directly to Model, CompDef (as I call it), or Group, As you mentioned elsewhere, the propensity is to want to write lines like, `model.Add(polygon)`, rather than `model.Entities.Add(polygon)`. To support this, I found myself writing oodles of forwarding methods that all just called methods of the same signature in the Entities objects, one each for Model, CompDef, and Group. That got tedious and _very_ error-prone. So, recognizing that three classes shared a lot of methods, I just elevated that into a parent class, and life has been a lot easier ever since.

It may _look_ odd, but I think it's not. Anyone with a lot of familiarity with the C API will be surprised by it, but the whole point of this layer is to avoid calling for such familiarity. A coder just wanting to build models and who knows nothing of the C API won't be a bit bothered by it. Also, it seems that the C API was written from a procedural, not OOP, perspective. "Subclassing," as we know it in OOP, isn't a relevant concept. If the C API had been designed as an OOP library, I think it is feasible that the "Entities" collection would have been named something else, and would serve as a parent class to anything that now has an Entities member.

As a design principal, recall that we agreed the users of this library should never see (and, therefore, never even have to know about) any C API details (like anything called SUEntitiesRef, and so on). Accordingly, I could have called the parent class of Model, CompDef, and Group, something like, "PolygonShepherd," and not had an "Entities" class at all.

Consider this: the coupling most of this library code has to SketchUp is in the constructors that take C API  references as their arguments (effectively, they are importers from SketchUp into the factory's DOM), and in the various Pack methods (effectively, they are exporters from the factory's DOM to a SketchUp model via the C API). Other than during import and export, there is no use made of the SketchUp C API at all. (That's why those constructors and the Pack methods are `internal`, not `public`. Client code never needs to see them.) To be sure I could import/export the structures needed for Unity work, I kept the factory's DOM similar to SketchUp's DOM. But, unlike the CsharpSketchUpAPI wrapper, the factory is not intended to be transliteral. Other than the coupling that exists to support import/export, the factory code uses its own DOM, with no particular need to have "instances" or "groups" _per se_ other than to facilitate export to a SketchUp file.

So, the short response is: Entities as a parent class got me out of three-times duplication of its "Add" methods that would otherwise have been necessary to avoid another level of additional dereferencing.

Will leave this issue open, while closing the others, so we can keep discussion in one place (don't want to imitate the Western hero who ends the story by jumping on his horse and riding off in all directions :) ).

**Comment by DanRathbun** [comment]
> I wanted client code to be able to add polygons directly to Model, CompDef (as I call it), or Group, As you mentioned elsewhere, the propensity is to want to write lines like, model.Add(polygon), rather than model.Entities.Add(polygon). To support this ...

It seems that you "picked your own poison". It was either just go with the actual complexity, or create your own complexity whilst trying to hide the collections from the OOP Layer users.

> I could have called the parent class of Model, CompDef, and Group, something like, "PolygonShepherd," and not had an "Entities" class at all.

I'd be less adverse to a different name. 
Perhaps `EntityGenerator` or `EntityFactory` or `EntityFabrication`.

**I'm not adverse to common reusable code**. 

In Ruby we have mixin library modules. ALL of the Ruby APIs model collection classes **`include`** the Ruby core [**`Enumerable`**](https://ruby-doc.org/core-2.5.5/Enumerable.html) module which adds in a truckload of methods, ie many search, filter and conversion iterators, etc. The mixin module is inserted into the mixee object's inheritance chain as a pseudo-superclass
(I guess the equivalent in C# is a static class? or perhaps an Interface?)

> So, the short response is: Entities as a parent class got me out of three-times duplication of its "Add" methods that would otherwise have been necessary to avoid another level of additional dereferencing.

Would it be better in C# to share functionality using an **Interface** instead of deriving from an abstract class?

> Also, it seems that the C API was written from a procedural, not OOP, perspective. "Subclassing," as we know it in OOP, isn't a relevant concept.

Of course because it is C. But, history ...

The SketchUp core is written in C++. The first SDK was a C++ SDK. (This might even have predated the Ruby API as it was wrapped with an `atlast` namespace.) 
But it wasn't so portable. And because of other considerations, they deprecated it (for ~ 6 years) and wrote the C API you are now using. (The C++ API and it's COM interface was finally removed with the 19.2 release.)

> If the C API had been designed as an OOP library, ...

C isn't OOP can't define native classes and objects. (It can define another OOP language like the Ruby Core quite well however.) 
But if you know the SketchUp DOM, you can see how the functions operate on objects and classes "under the hood". The functions are organzied and named with class prefixes. (Some of them use a differing prefix than the Ruby API. Ex: `SUPoint3D` as opposed to `Geom::Point3d`. Note: All of the classes in the `Geom` module are not model objects. They're virtual helpers classes. Ie, models don't have points, they have geometric `SUVertex`/`Sketchup::Vertex` instances This is why the former is in the `Geom` namespace, and the latter in the `Sketchup` namespace in the OOP Ruby API.)

> ... I think it is feasible that the "Entities" collection would have been named something else, and would serve as a parent class to anything that now has an Entities member.

This wouldn't be true, because the internal core is C++ and it's model and component definitions, each instantiate a singleton instance of an "Entities" C++ collection. It may not be named exactly "Entities" in the core C++ code, but "Entities" is the exposed name given the class in the Ruby API. (I've said in the past that I dislike plural names for collections and prefer the suffixes "Set" and "List".)

But what it's name is doesn't matter in other APIs, even in the 12+ year old Ruby API.

What does matter is that the real scenario is that: a "semi-static" instance object S who "owns" it's particular singleton instance of a collection class E that is also instantiated by other objects C (whose existence will come and go in another of the "semi-static" S object's singleton collections D,) each of whom also "owns" it's particular singleton instance of the collection class E.
What's more is that members of the instances of collection classes E and D, have properties (fields) that reference members of several other singleton collections (L, M, T and P) of "semi-static" instance S. 
Further complicating is that every one of the aforementioned collections, and every one of it's individual members can have a collection of A members (of which each member can have a collection of A members.)
And this is overly simplified. There are a few more model collections not mentioned (that you've not _yet wrapped_,) and a bunch more application collections (but they're not saved into the `.skp` file.)

Having S a subclass of E, whilst it contains a collection instance of E, and a collection D whose members are also a subclass of E, whilst they each contain a collection instance of E, ... would be not only weird, but likely **unworkable**. (Vicious reference circles )
**They did not do this, and neither did you.** 

You created a common code object (that is NOT a collection class, but given a collection class' name,) in order to facilitate the common use of functionality (via inheritance.)

There's nothing wrong with linking in common code by whatever means. They might even have common core entity factory functionality superclasses in the internal C++ code that we don't know about (or need to care about.)

>It may look odd, but I think it's not. Anyone with a lot of familiarity with the C API will be surprised by it, but the whole point of this layer is to avoid calling for such familiarity. A coder just wanting to build models and who knows nothing of the C API won't be a bit bothered by it.

Serious developers **are** bothered when things are hidden from them, or not exposed to APIs.

Hiding collection access from "live" extension developers will not work. There are many extensions that need to operate directly on collections. Many extensions would not be implementable otherwise.

----
----

> (That's why those constructors and the Pack methods are internal, not public. Client code never needs to see them.)

I meant in your OOP Layer
_from `"SketchUpFactoryExamples/Examples/GroupQuad.cs"`_
```c#
    Model model = new Model();

    // Create the group, with a name.
    Group group = new Group("I am a group name");

    // Add the group to the model.
    model.Add(group);
```
A slight semantic difference in full OOP is that the root object (a `Model` instance in this case,) is requested from "the engine". In the "live" use (Ruby API as example) it's the `Sketchup` module object representing the application that is "the engine".
Note that Ruby's `Module` _class serves as_ both the superclass of class `Class` _and it's instances as namespaces_. The `Module` class' methods are inherited by all classes. A module instance can have instance and module methods. 
So the `Sketchup` module does both. It has application level module methods (not needed in standalone use,) and serves as a namespace for the API's classes of things that go into a model, work with a model, or control the application and it's extensions.)
**Ex:** `Sketchup.file_new`, `SketchUp.file_open`.
But I don't see direct equivalent to modules in C#. (Maybe a `static` or `sealed` class?)

Anyway, you have used the overloaded `Model` class constructor for the two scenarios.
`Model model =` ... either ... `new Model()` ... or ... `new Model("C:/some/file/path/model,skp")`

Can't complain too much about this especially in a standalone use. The model needs to be instantiated somehow. If a `ModelFactory` object was available you could go full OOP ...
```c#
    Model model = ModelFactory.New()
```
... or ...
```c#
    Model model = ModelFactory.Open("C:/some/file/path/model.skp")
```
This would allow a path to be passed to the model's path field, to be used for any subsequent saving operations. When directly calling only one of the 2 overloaded `Model` constructors, code cannot do this (pass initial field values for the blank model.)

I suppose you could also define class factory methods upon the existing `Model` class that in turn call the constructors, and then coders could use one or the other means... yes?
```c#
    Model model = Model.NewFile()
```
... or ...
```c#
    Model model = Model.OpenFile("C:/some/file/path/model.skp")
```
This would work for standalone use, but _perhaps_ not for internal to the SketchUp application. Only the Mac edition has a MDI. The Windows edition can only open 1 model **for edit**. Using the SDK, I suppose we can presently open or create another model in memory, to say read from the "active model" and write the memory model.

You also have an unattached group possibility. In OOP we'd be more used to ...
```c#
    // Create the group, with a name.
    Group group = model.AddGroup("I am a group name");
```
... and the "AddGroup" factory method does the constructor call and **ensures** that it is attached to the model object whose's "AddGroup" factory method was called, then returns the `Group` instance object.
This would also open the avenue of parameter overrides for the group's fields: layer, material, transformation or description. (Instead of forcing a series of setter statements following the Add call.)

**Comment by DanRathbun** [comment]
> My goal is to provide something that benefits from, but does not require, knowledge of the SketchUp DOM. Since I am using it myself, it has to be simple, clean, and easy to reacquaint myself with after time away from it. ... So, your assessment matches my objective. A concise implementation, but enough to get the job done.

**So the jist I take away is: It is a simple quick and dirty OOP layer for making Unity game models.
It isn't designed for and probably won't be for mainstream big architectural models and "live" SketchUp extensions.**

It has the "Goldielocks syndrome": Long time SketchUp API DOM coders may think it too simple. Newbies may say "it's just right."

That's fine. Things have there uses.

**It also serves as a good example of what can be done in C#.**

**Comment by stevensrmiller** [comment]
Closing this so it isn't just here forever.
